File: MarcoPlaatsAPI/Helpers/RestfulController.cs
namespace MarcoPlaatsAPI
{
    using System;
    using Nancy;

    public abstract class RestfulController : NancyModule
    {
        public RestfulController(string prefix)
        {
            prefix = "/" + prefix + "/";

            Get(prefix, _ => this.Index());
            Get(prefix + "{id}", _ => this.Show(_.id));
            Post(prefix, _ => this.Create());
            Put(prefix + "{id}", _ => this.Update(_.id));
            Delete(prefix + "{id}", _ => this.Delete(_.id));
        }

        public abstract string Index();
        public abstract string Create();
        public abstract string Show(string id);
        public abstract string Update(string id);
        public abstract string Delete(string id);
    }
}

File: API/Controllers/RestfulController.js
module.exports = class RestfulController {
    /**
     * @property {Model} model
     * @param {function} model
     */
    constructor(model) {
        this.model = new model();
    }

    index(req, res, next) {
        this.model.params = req.params;

        this.model.all()
            .then(results => {
                res.send(results)
            })
            .catch(next)
    }

    create(req, res, next) {
        this.model.params = req.params;
        this.model.document = req.body;

        this.model.insert()
            .then(() => {
                res.statusCode = 201;
                res.send(this.model.document);
            })
            .catch(next)
    }

    show(req, res, next) {
        this.model.params = req.params;

        this.model.findById(req.params.id)
            .then(() => {
                res.send(this.model.document)
            })
            .catch(next)
    }

    update(req, res, next) {
        this.model.params = req.params;

        this.model.update()
            .then(() => {
                res.statusCode = 204;
                res.end()
            })
            .catch(next)
    }

    destroy(req, res, next) {
        this.model.params = req.params;

        this.model.findById(req.params.id)
            .then(() => {
                if (this.model.document == null) {
                    res.statusCode = 404;
                    res.end();
                    return
                }

                this.model.destroy()
                    .then(() => {
                        res.statusCode = 204;
                        res.end()
                    })
                    .catch(next)
            })
            .catch(next)
    }
};
